app/services/sync_provider_from_find.rb

```diff
@@ -112,7 +95,48 @@ def create_or_update_course(find_course)
       end
     end
 
-    course
+    handle_course_options_with_invalid_sites(course, find_course)
```

Comment:
When discussing this with @tijmenb originally, he noted that study modes were potentially another source of 'invalid' course options. That is, when a course study mode changes from `both_full_time_and_part_time` to `part_time`, let's say, the course options for `full_time` will no longer get picked up by the `CourseOption.find_or_create_by` above. From what I gather, the course study mode can change depending on the availability of vacancies for a given study mode at any given time.

I'm not sure these course options should be handled with the same 'delete or warn' approach taken in `handle_course_options_with_invalid_sites` below. Reasons:

- It's likely that a higher number of these course options will be part of an ApplicationChoice, and there's little value in warning about them because these are completely legitimate picks (whereas if the course site list changes, we probably want to be aware of any in-progress applications that have selected a site that no longer supports the chosen course).
- Deleting course options that aren't part of an ApplicationChoice is feasible, but it doesn't seem necessary - if a course's study mode changes to either FT or PT, we don't give the candidate a choice during course selection anyway.

Possible alternative:

When deciding what the study modes are further up in this method, rather than use `course.both_study_modes_available?`, we do something slightly more involved:

- Does this course have existing course options?
- If so, determine the `study_modes` array from the unique set of study modes among those course options.
- If not, just do the existing check.

This lets us operate against all CourseOptions created for a course historically, irrespective of the current course study mode returned from Find. We then set the vacancy status for each option, which will correspond with the study mode of the course (ie - if the course study mode is `full_time`, all of the `part_time` course options will have `no_vacancies`). WDYT?

Reply:
This sounds sensible to me. My understanding of the issue:

1. We pull a Course (C) with `full_time_or_part_time` `study_mode`.
2. We create options (O1 and O2) with `full_time` and `part_time` `study_modes` respectively.
3. The `study_mode` of C changes to `full_time` because there are no longer any part time places.
4. On next sync, we iterate over the `study_mode`s of C to find the associated `CourseOptions`. We find O2 because it's part time, but don't find O1. Therefore we don't update O1's vacancy status, which is bad because it remains available for applications when it's actually full.

If this is right, the proposed solution sounds sensible.

Reply:
@duncanjbrown that's correct, with particular emphasis on "we iterate over the study_modes of C". Specifically, we determine what the study modes are based on the value of a single enum field on C. We should instead be deriving the study modes from existing options, if they exist.

I'll put the study mode changes in a separate PR, leaving this one focused on the site-related changes.